File: src/war3map/UnitControllers/BasicUnitController.lua
local ____exports = {}
local __TSTL_Point = require("war3map.Generic.Point")
local Point = __TSTL_Point.Point
local __TSTL_ExtensionFunctions = require("war3map.ExtensionFunctions")
local IsWalkable = __TSTL_ExtensionFunctions.IsWalkable
____exports.BasicUnitController = {}
local BasicUnitController = ____exports.BasicUnitController
BasicUnitController.name = "BasicUnitController"
BasicUnitController.__index = BasicUnitController
BasicUnitController.prototype = {}
BasicUnitController.prototype.__index = BasicUnitController.prototype
BasicUnitController.prototype.constructor = BasicUnitController
function BasicUnitController.new(...)
    local self = setmetatable({}, BasicUnitController.prototype)
    self:____constructor(...)
    return self
end
function BasicUnitController.prototype.____constructor(self, targetUnit, heroUnit)
    self.currentDirection = bj_UNIT_FACING
    self.targetUnit = targetUnit
    self.walkAnimationIndex = heroUnit.walkAnimationIndex
    self.heroUnit = heroUnit
end
function BasicUnitController.prototype.moveUnit(self, toDirection)
    local speed = GetUnitMoveSpeed(self.targetUnit) / 100
    self.currentDirection = self:rotateToPoint(toDirection, 180)
    local lastLoc = Point:fromLocationClean(GetUnitLoc(self.targetUnit))
    local currLoc = lastLoc:polarProject(speed, self.currentDirection)
    if IsWalkable(nil, currLoc.x, lastLoc.y) then
        SetUnitX(self.targetUnit, currLoc.x)
    end
    if IsWalkable(nil, lastLoc.x, currLoc.y) then
        SetUnitY(self.targetUnit, currLoc.y)
    end
    SetUnitFacing(self.targetUnit, self.currentDirection)
end
function BasicUnitController.prototype.getWalkAnimationIndex(self)
    return self.walkAnimationIndex
end
function BasicUnitController.prototype.rotateToPoint(self, toDir, turnSpeed)
    local result = toDir - self.currentDirection
    while result > 180 do
        result = result - 360
    end
    while result < -180 do
        result = result + 360
    end
    local turnDir = result
    if turnDir < -turnSpeed then
        turnDir = -turnSpeed
    end
    if turnDir > turnSpeed then
        turnDir = turnSpeed
    end
    return (self.currentDirection + turnDir) % 360
end
return ____exports

File: war3map/UnitControllers/BasicUnitController.ts
import {IUnitController} from "./IUnitController";
import {HeroUnit} from "./HeroUnit";

export class BasicUnitController implements IUnitController {
    private readonly targetUnit: unit;
    private readonly walkAnimationIndex: number;
    private currentDirection: number = bj_UNIT_FACING;
    public heroUnit: HeroUnit;


    constructor(targetUnit: unit, heroUnit: HeroUnit) {
        this.targetUnit = targetUnit;
        this.walkAnimationIndex = heroUnit.walkAnimationIndex;
        this.heroUnit = heroUnit;

    }

    moveUnit(toDirection: number) {
        let speed = GetUnitMoveSpeed(this.targetUnit) / 100;
        this.currentDirection = this.rotateToPoint(this.currentDirection, toDirection, 180);
        let currLoc = GetUnitLoc(this.targetUnit);
        currLoc = PolarProjectionBJ(currLoc, speed, this.currentDirection);
        SetUnitPositionLoc(this.targetUnit, currLoc);
        SetUnitFacing(this.targetUnit, this.currentDirection);
        RemoveLocation(currLoc);
    }

    getWalkAnimationIndex(): number {
        return this.walkAnimationIndex;
    }


    private rotateToPoint(fromDir: number, toDir: number, turnSpeed: number) {
        let result = toDir - fromDir;
        while (result > 180) {
            result -= 360
        }
        while (result < -180) {
            result += 360
        }
        let turnDir = result;

        if (turnDir < -turnSpeed) {
            turnDir = -turnSpeed
        }
        if (turnDir > turnSpeed) {
            turnDir = turnSpeed
        }

        return (fromDir + turnDir) % 360;
    }
}